### ISSUE TITLE
File Appending Failing With Identical Schema

### ISSUE THREAD

**Issue by houseofcat** [aloneguid/parquet-dotnet#134]
**Version:** Parquet.Net v3.9.1

**Runtime Version:** NET5.0

**OS:** Windows

#### Expected behavior

Appending file with the exact same Schema object that created the file to begin with should keep working when switching to appendation.

#### Actual behavior

Parquet is reporting a data inconsistency on type int16, the column `OrderQty` of the Adventureworks 2019 database on one table: `[Sales].[SalesOrderDetail]`

### Steps to reproduce the behavior

1. AdventureWorks 2019, SqlServer 2019 Developer Edition
2. SELECT * FROM [Sales].[SalesOrderDetail]
3. Loop through DataReader making a Parquet Table filled with Parquet Rows.
4. Write Parquet Table to ParquetRowGroup with Parquet Writer (all to FileStream).
5. If there is more data in Reader after the first past, write again to the same FileNamePath but with append true and a new FileStream.

#### Code snippet reproducing the behavior

```csharp
while (!reader.IsClosed)
{
    using var fileStream = File.Open(fileNamePath, FileMode.OpenOrCreate);
    var currentFileSize = new FileInfo(fileNamePath).Length;

    // This is the file size after the first RowGroup is
    // written to file (with compression included).
    if (currentFileSize > 0 && postCompressedRowGroupSize == 0)
    { postCompressedRowGroupSize = currentFileSize; }

    if (maxFileSize > currentFileSize)
    {
        var remainingFileSize = maxFileSize - currentFileSize;

        // File size is good enough check.
        //
        // If we proceed past this point, postCompressedRowGroupSize
        // will be applied and write essentially up full RowGroup to file.
        // If the RowGroup is sufficiently large enough, this could
        // drastically alter the file size.
        if (remainingFileSize < 8096)
        { return; }

        // Prevents excessive RowGroupPartitions at the end of the file
        // by setting the maximum table size to the first row page group
        // compressed size. This should allow us to create one
        // final partition.
        remainingFileSize = remainingFileSize > postCompressedRowGroupSize
            ? remainingFileSize
            : postCompressedRowGroupSize;

        var table = BuildTable(reader, schema, remainingFileSize, maxRowGroupSize);

        using var parquetWriter = new ParquetWriter(schema, fileStream, append: appendFile)
        {
            CompressionMethod = compressionMethod,
            CompressionLevel = compressionLevel,
        };

        using ParquetRowGroupWriter groupWriter = parquetWriter.CreateRowGroup();

        groupWriter.Write(table);

        // If the reader is not closed, we need to switch to append file.
        if (!reader.IsClosed) { appendFile = true; }
    }
    // This file is at maximum size. Reader still open indicates more to write
    // but it's time for a new file.
    else
    { return; }
}
```

Schema @ First Write (Append == False)
![image](https://user-images.githubusercontent.com/1414138/141557424-4d20319b-3cc5-41c6-9f27-48250203cf0b.png)

Schema @ Second Write (Append == True)
![image](https://user-images.githubusercontent.com/1414138/141559342-a3060a08-d4e9-4f6a-a493-5a3ff58edaa2.png)

Exception
![image](https://user-images.githubusercontent.com/1414138/141560909-a9156ac4-bb45-4e6c-b640-34c9c3c3fa0e.png)

**Comment by houseofcat** [comment]
Tracked the issue down.

Reproducible from a hijacked Schema UnitTest that shows the Schema.Equals() is passing but not the Validation of Schemas in Parquet Writer code.

```
[Fact]
public void Schemas_identical_WithGroupWriter()
{
   var schema1 = new Schema(new DataField<int>("id"), new DataField("quantity", DataType.Int16, hasNulls: true));
   var schema2 = new Schema(new DataField<int>("id"), new DataField("quantity", DataType.Int16, hasNulls: true));

   Assert.True(schema1.Equals(schema2));

   int count = 0;
   bool append = false;
   while (count < 3)
   {
      if (count == 1) { append = true; }

      Int32 intValue = 1;
      Int16? shortValue = 2;
      var table = new Table(schema1)
      {
         new Row(
            new object[] { intValue, shortValue })
      };

      using FileStream fileStream = System.IO.File.Open("./test.txt", FileMode.OpenOrCreate);
      using var parquetWriter = new ParquetWriter(schema1, fileStream, append: append)
      {
         CompressionMethod = CompressionMethod.Snappy,
         CompressionLevel = -1,
      };

      using ParquetRowGroupWriter groupWriter = parquetWriter.CreateRowGroup();

      groupWriter.Write(table);
      count++;
   }
}
```

**Comment by houseofcat** [comment]
The code to blame is located here in the ParquetWriter.cs

Specifically the `Schema.Equals` code.
```csharp
private void ValidateSchemasCompatible(ThriftFooter footer, Schema schema)
{
   Schema existingSchema = footer.CreateModelSchema(_formatOptions);

   if (!schema.Equals(existingSchema))
   {
      string reason = schema.GetNotEqualsMessage(existingSchema, "appending", "existing");
      throw new ParquetException($"passed schema does not match existing file schema, reason: {reason}");
   }
}
```

There are two complex C# objects using the basic System.Object `Equals`, both boxed as `Field` type but are really `DataFields`.

This is an uncommon practice in C# to rely on System.Object equals of complex object equality. Unfortunately, from the outside there is a complexity here and could theoretically make a breaking change. I will submit a simple PR you can modify to your liking.

**Comment by houseofcat** [comment]
Should be fixed by PR #135
